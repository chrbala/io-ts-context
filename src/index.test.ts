import { pipe } from 'fp-ts/function';
import * as D from 'io-ts/Decoder';
import * as C from './';
import * as E from 'fp-ts/Either';

const unsafeParse = <T, C>(
  decoder: C.ContextualDecoder<unknown, T, C>,
  context: C,
  value: unknown
) => {
  const result = decoder(context).decode(value);
  if (E.isLeft(result)) throw new Error(D.draw(result.left));

  return result.right;
};

describe('literal', () => {
  const literal = C.literal('A', 'B', 'C');

  test('valid', () => {
    const actual = unsafeParse(literal, null, 'A');
    const expected = 'A';
    expect(actual).toEqual(expected);
  });

  test('invalid', () => {
    expect(() =>
      unsafeParse(literal, null, 'D')
    ).toThrowErrorMatchingSnapshot();
  });
});

describe('number', () => {
  type Context = {
    validateNumber: (num: number) => null | string;
  };
  const num = pipe(
    C.number,
    C.parse((s, c: Context) => {
      const error = c.validateNumber(s);
      return error ? D.failure(s, error) : D.success(s);
    })
  );

  const context: Context = {
    validateNumber: num => (num > 0 ? null : 'positive'),
  };

  test('valid', () => {
    const actual = unsafeParse(num, context, 5);
    const expected = 5;

    expect(actual).toEqual(expected);
  });

  test('invalid', () => {
    expect(() => unsafeParse(num, context, -5)).toThrowErrorMatchingSnapshot();
  });
});

describe('string', () => {
  type Context = {
    validateString: (str: string) => string | null;
  };
  const num = pipe(
    C.string,
    C.parse((s, c: Context) => {
      const error = c.validateString(s);
      return error ? D.failure(s, error) : D.success(s);
    })
  );

  const context: Context = {
    validateString: str => (str.match(/^[A-Z]*$/) ? null : 'capitalized'),
  };

  test('valid', () => {
    const actual = unsafeParse(num, context, 'CAPS');
    const expected = 'CAPS';

    expect(actual).toEqual(expected);
  });

  test('invalid', () => {
    expect(() =>
      unsafeParse(num, context, 'lowercase')
    ).toThrowErrorMatchingSnapshot();
  });
});

describe('type', () => {
  type Context = {
    validateNumber: (num: number) => null | string;
  };
  const num = pipe(
    C.number,
    C.parse((s, c: Context) => {
      const error = c.validateNumber(s);
      return error ? D.failure(s, error) : D.success(s);
    })
  );

  const context: Context = {
    validateNumber: num => (num > 0 ? null : 'positive'),
  };

  const obj = C.type<Context>()({
    num,
  });

  test('valid', () => {
    const actual = unsafeParse(obj, context, { num: 5 });
    const expected = { num: 5 };

    expect(actual).toEqual(expected);
  });

  test('invalid', () => {
    expect(() =>
      unsafeParse(obj, context, { num: -5 })
    ).toThrowErrorMatchingSnapshot();
  });
});
