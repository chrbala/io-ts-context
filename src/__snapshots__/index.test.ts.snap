// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`literal invalid 1`] = `"cannot decode \\"D\\", should be \\"A\\" | \\"B\\" | \\"C\\""`;

exports[`number invalid 1`] = `"cannot decode -5, should be positive"`;

exports[`string invalid 1`] = `"cannot decode \\"lowercase\\", should be capitalized"`;

exports[`type invalid 1`] = `
"required property \\"num\\"
└─ cannot decode -5, should be positive"
`;
